lib/vanagon/optparse.rb

```diff
@@ -2,40 +2,86 @@
-        :verbose => ['-v', '--verbose', 'Verbose output (does nothing)'],
-        :only_build => ["--only-build COMPONENT,COMPONENT,...", Array, 'Only build this array of components']
-      }.freeze
-
-    def initialize(banner, options = [])
-      @options = Hash[options.map { |v| [v, nil] }]
-      @optparse = OptionParser.new do |opts|
+    def initialize(banner, symbols = []) # rubocop:disable Metrics/AbcSize
+      ## symbols array kept for backward compatibility but ignored
+
+      @options = Hash.new
+      @options[:preserve] = :never
+
+      @option_parser = OptionParser.new do |opts| # rubocop:disable Metrics/BlockLength
         opts.banner = banner
+        opts.separator ""
+        opts.separator "Options:"
+
+        opts.on("-h",
+                "--help",
+                "Display help") do
+          $stdout.puts opts
+          exit 1
+        end
+
+        opts.on("-v",
+                "--verbose",
+                "Verbose output (does nothing)") do |verbose|
+          @options[:verbose] = verbose
+        end
+
+        opts.on("-w DIRECTORY",
+                "--workdir DIRECTORY",
+                "Working directory on the local host (defaults to calling mktemp)") do |workdir|
+          @options[:workdir] = workdir
+        end
+
+        opts.on("-r DIRECTORY",
+                "--remote-workdir DIRECTORY",
+                "Working directory on the remote host (defaults to calling mktemp)") do |remote|
+          @options[:"remote-workdir"] = remote
+        end
+
+        opts.on("-c DIRECTORY",
+                "--configdir DIRECTORY",
+                "Configuration directory (defaults to $CWD/configs)") do |configuration_directory|
+          @options[:configdir] = configuration_directory
+        end
 
-        FLAGS.each_pair do |name, args|
-          if @options.include?(name)
-            opts.on(*args) do |value|
-              @options[name] = value
-            end
+        opts.on("-t HOST",
+                "--target HOST",
+                "Name of target machine for build and packaging (defaults to requesting from the pooler)") do |target|
+          @options[:target] = target
+        end
+
+        opts.on("-e ENGINE",
+                "--engine ENGINE",
+                "Custom engine to use [base, local, docker, pooler] (defaults to \"pooler\")") do |engine|
+          @options[:engine] = engine
+        end
+
+        opts.on("--skipcheck",
+                "Skip the \"check\" stage when building components") do |skipcheck|
+          @options[:skipcheck] = skipcheck
+        end
+
+        opts.on("-p [RULE]",
+                "--preserve [RULE]",
+                ["never", "on-failure", "always"],
+                "Rule for VM preservation. [never, on-failure, always]") do |rule|
+          if rule.nil?
+            @options[:preserve] = :always
+          else
+            @options[:preserve] = rule
```

Comment:
So, question. Above, we're checking the contents of this option against symbols, but there's nothing here that enforces that this is a symbol. Is this me being overly concerned? Or is there something here about OptionParser that I'm missing that handles that?

Also, on line 9 you have `@options[:preserve] = :never`. It seems like we have conflicting defaults here.

Reply:
First point: agreed, this could be cleaner. It functions as-is but it's playing it too loose with Ruby conveniences.

Second point: The defaulting behavior is as we agreed:

- With no --preserve argument: preserve is set to 'never'
- With --preserve but no argument: preserve is set to 'always'
- With --preserve=\<argument\>: preserve is set to \<argument\>

Reply:
Oh, derp, that makes perfect sense. Thank you!

And to the symbol check, there are a lot of places in our code where we play it fast and loose with Ruby conveniences. If this is working currently, then I suppose there's no reason to be too concerned until this starts to fail.